payments/payment-methods/adyen/src/main/java/cm/aptoide/pt/payment_method/adyen/repository/AdyenV2Repository.kt

```diff
@@ -28,46 +35,161 @@ internal class AdyenV2RepositoryImpl @Inject constructor(
     walletAddress: String,
     priceValue: String,
     priceCurrency: String,
-  ): PaymentMethodDetailsData {
+  ): PaymentMethodDetailsData = try {
     val response = adyenV2Api.getPaymentMethodDetails(
       ewt = "Bearer $ewt",
       walletAddress = walletAddress,
       priceValue = priceValue,
       priceCurrency = priceCurrency,
     )
 
-    return PaymentMethodDetailsData(
+    PaymentMethodDetailsData(
       price = response.price.value.toDouble(),
       currency = response.price.currency,
       json = JSONObject(response.payment.toString())
     )
+  } catch (e: Throwable) {
+    throw extractAdyenException(e)
   }
 
   override suspend fun createTransaction(
     ewt: String,
     walletAddress: String,
     paymentDetails: PaymentDetails,
-  ): TransactionResponse = adyenV2Api.createTransaction(
-    ewt = "Bearer $ewt",
-    walletAddress = walletAddress,
-    paymentDetails = paymentDetails
-  )
+  ): TransactionResponse {
+    val response = try {
+      adyenV2Api.createTransaction(
+        ewt = "Bearer $ewt",
+        walletAddress = walletAddress,
+        paymentDetails = paymentDetails
+      )
+    } catch (e: Throwable) {
+      throw extractAdyenException(e)
+    }
+
+    if (response.status == FAILED || response.status == CANCELED
+      || response.status == DUPLICATED
+    ) {
+      val refusalReasonCode = response.payment?.refusalReasonCode
+      val refusalReason = response.payment?.refusalReason
+
+      if (refusalReasonCode != null && refusalReason != null) {
+        throw mapAdyenRefusalCode(refusalReasonCode, refusalReason)
+      } else {
+        throw AdyenRefusalException()
+      }
+    }
+
+    return response
+  }
 
   override suspend fun submitActionResult(
     ewt: String,
     walletAddress: String,
     uid: String,
     paymentData: String?,
     paymentDetails: JSONObject?,
-  ): TransactionResponse = adyenV2Api.submitActionResult(
-    ewt = "Bearer $ewt",
-    walletAddress = walletAddress,
-    uid = uid,
-    request = AdyenPayment(
-      data = paymentData,
-      details = paymentDetails.toJsonObject()
-    )
-  )
+  ): TransactionResponse {
+    val response = try {
+      adyenV2Api.submitActionResult(
+        ewt = "Bearer $ewt",
+        walletAddress = walletAddress,
+        uid = uid,
+        request = AdyenPayment(
+          data = paymentData,
+          details = paymentDetails.toJsonObject()
+        )
+      )
+    } catch (e: Throwable) {
+      throw extractAdyenException(e)
+    }
+
+    if (response.status == FAILED || response.status == CANCELED
```

Comment:
Let's encopsulate this if into `mapAdyenRefusalCode` and make that return optional `AdyenRefusalException`
So the usage could look like so (avoiding duplicated code):
`response.mapAdyenRefusalCode()?.let{ throw it }`